godot_4{,-mono}: 4.3-stable -> 4.4-stable

corngood opened the pull request:
<!--
^ Please summarise the changes you have done and explain why they are necessary here ^

For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#linking-nixos-module-tests-to-a-package) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [25.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2505.section.md) (or backporting [24.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) and [25.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2505.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#reviewing-contributions
-->

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

rummik commented:
I think the only thing missing is the hash in `godot_4-export-templates` needs to be updated, as it references `godot_4`'s version.

It also looks like some packages break when using Godot 4.4 export templates, like `pixelorama`

corngood commented:
This is totally broken at the moment.  The update script doesn't work with the `commitHash` thing.

corngood commented:
## `nixpkgs-review` result

Generated using [`nixpkgs-review`](https://github.com/Mic92/nixpkgs-review).

Command: `nixpkgs-review pr 387031`

---
### `x86_64-linux`
<details>
  <summary>:white_check_mark: 9 packages built:</summary>
  <ul>
    <li>find-billy</li>
    <li>godot_4</li>
    <li>godot_4-mono</li>
    <li>godot_4-mono.debug</li>
    <li>godot_4-mono.man</li>
    <li>godot_4.debug</li>
    <li>godot_4.man</li>
    <li>opengamepadui</li>
    <li>pixelorama</li>
  </ul>
</details>

corngood commented:
The removal of `commitHash` will probably undo `f37c9c60d2c831e1b1998bd3bd31fab7833ff7ad`, so I'll try to find a way to fix it.

drupol commented on pkgs/by-name/go/godot_4/package.nix:
```diff
@@ -63,14 +63,21 @@ let
       repo = "godot";
-      rev = commitHash;
-      hash = "sha256-v2lBD3GEL8CoIwBl3UoLam0dJxkLGX0oneH6DiWkEsM=";
+      rev = version;
```
```suggestion
      tag = version;
```

corngood replied:
Done.  I guess the difference here is that `tag` will match the tag explicitly `refs/tags/${tag}` rather than possibly being ambiguous?

corngood commented:
Marking this as draft because there's something wrong with `godot_4.src` (failing builds on borg).  It seems to be non-deterministic???

corngood commented:
I've added the backport label because it's a minor version update and the existing package is marked insecure.  Does anyone have any concerns about that?

rummik commented:
I believe `godot_4-export-templates` is supposed to be updated in tandem with `godot_4`, as it references `godot_4.version` in its construction 

In 24.11 it looks like `pixelorama` is the only package that depends on `godot_4-export-templates`, and that version of Pixelorama breaks with Godot 4.4's templates

adamtulinius commented:
> I believe `godot_4-export-templates` is supposed to be updated in tandem with `godot_4`, as it references `godot_4.version` in its construction

Since godot are introducing breaking changes on minor releases (see https://docs.godotengine.org/en/4.4/tutorials/migrating/upgrading_to_godot_4.4.html#breaking-changes), maybe the export templates needs to exist in multiple minor versions as well in nixpkgs?

rummik commented:
It seems like it might also be off-putting to be working on a project in 4.3 and suddenly have the package become 4.4 while on a stable nixpkgs release

corngood commented:
Fair enough.  I'll remove the backport label for now. 

> I believe godot_4-export-templates is supposed to be updated in tandem with godot_4, as it references godot_4.version in its construction

This seems pretty fragile:

```
  url = "https://github.com/godotengine/godot/releases/download/${godot_4.version}/Godot_v${godot_4.version}_export_templates.tpz";
  hash = "sha256-XRnKii+eexIkbGf7bqc42SR0NBULFvgMdOpSRNNk6kg=";
```

I'll update it, and I think I can probably add a version test to it.

corngood committed:
godot_4-export-templates: add version to derivation

corngood commented:
## `nixpkgs-review` result

Generated using [`nixpkgs-review`](https://github.com/Mic92/nixpkgs-review).

Command: `nixpkgs-review pr 387031`

---
### `x86_64-linux`
<details>
  <summary>:white_check_mark: 10 packages built:</summary>
  <ul>
    <li>find-billy</li>
    <li>godot_4</li>
    <li>godot_4-export-templates</li>
    <li>godot_4-mono</li>
    <li>godot_4-mono.debug</li>
    <li>godot_4-mono.man</li>
    <li>godot_4.debug</li>
    <li>godot_4.man</li>
    <li>opengamepadui</li>
    <li>pixelorama</li>
  </ul>
</details>

corngood commented:
> In 24.11 it looks like pixelorama is the only package that depends on godot_4-export-templates, and that version of Pixelorama breaks with Godot 4.4's templates

This seems to be the case in master as well.  It just exits after the splash screen without any obvious error.  I'm not sure if this is expected with 4.4, or if our build of 4.4 is broken somehow.

rummik commented:
What's really weird is changing `--export-release` to `--export-debug` in the `buildPhase` hook for the `pixelorama` package doesn't crash

corngood commented:
> What's really weird is changing --export-release to --export-debug in the buildPhase hook for the pixelorama package doesn't crash

Yeah, that's annoying.  It's not the .pck file either, because it doesn't crash if run with the debug binary or godot4 itself.  The export template executables don't seem to have symbols unfortunately, and haven't been able to track them down.

The first print that happens after the release build crashes is:

```
SCRIPT ERROR: Cannot call method 'set_item_text' on a null value.
          at: _update_file_menu_buttons (res://src/UI/TopMenuContainer/TopMenuContainer.gd:133)
```

Any chance godot disables null checks in release in such a way that it'll just SIGSEGV?

corngood commented:
I tried some test projects on the release player that pixelorama uses, and they seem fine, so I don't think there's a fundamental problem with our build of 4.4.

Some thoughts:

- we should probably package 4.3 for the projects that use them
- we should really be building the export templates (player binary) rather than pulling binary blobs, especially where it's used in nixpkgs (we can provide the binary one separately)

corngood committed:
godot: move all packages into pkgs/development/tools/godot

corngood committed:
godot_4: compute commitHash in fetcher

corngood committed:
godot_4-mono: remove dotnet-sdk_6 dependency

corngood committed:
godot: support multiple versions

corngood committed:
godot_4_4: init at 4.4-stable

corngood committed:
opengamepadui: use godot 4.3 explicitly

corngood committed:
pixelorama: use godot 4.3 explicitly

corngood committed:
godot_4: default to godot 4.4

corngood commented:
## `nixpkgs-review` result

Generated using [`nixpkgs-review`](https://github.com/Mic92/nixpkgs-review).

Command: `nixpkgs-review pr 387031`

---
### `x86_64-linux`
<details>
  <summary>:white_check_mark: 17 packages built:</summary>
  <ul>
    <li>find-billy</li>
    <li>godot (godot_4, godot_4_4)</li>
    <li>godot-export-templates (godot_4-export-templates, godot_4_4-export-templates)</li>
    <li>godot-mono (godot_4-mono, godot_4_4-mono)</li>
    <li>godot-mono.debug (godot_4-mono.debug, godot_4_4-mono.debug)</li>
    <li>godot-mono.man (godot_4-mono.man, godot_4_4-mono.man)</li>
    <li>godot.debug (godot_4.debug, godot_4_4.debug)</li>
    <li>godot.man (godot_4.man, godot_4_4.man)</li>
    <li>godot_4_3</li>
    <li>godot_4_3-export-templates</li>
    <li>godot_4_3-mono</li>
    <li>godot_4_3-mono.debug</li>
    <li>godot_4_3-mono.man</li>
    <li>godot_4_3.debug</li>
    <li>godot_4_3.man</li>
    <li>opengamepadui</li>
    <li>pixelorama</li>
  </ul>
</details>

corngood commented:
This now supports multiple versions and includes both 4.3/4.4.

I'm not a big fan of how big this is getting:

```
  godotPackages = callPackage ../development/tools/godot { };

  inherit (godotPackages)
    godot3
    godot3-export-templates
    godot3-headless
    godot3-debug-server
    godot3-server
    godot3-mono
    godot3-mono-export-templates
    godot3-mono-headless
    godot3-mono-debug-server
    godot3-mono-server
    ;

  godot_4_3 = godotPackages.godot_4_3.godot;
  godot_4_3-mono = godotPackages.godot_4_3.godot-mono;
  godot_4_3-export-templates = godotPackages.godot_4_3.godot.export-templates;
  godot_4_4 = godotPackages.godot_4_4.godot;
  godot_4_4-mono = godotPackages.godot_4_4.godot-mono;
  godot_4_4-export-templates = godotPackages.godot_4_4.godot.export-templates;
  godot_4 = godotPackages.godot_4.godot;
  godot_4-mono = godotPackages.godot_4.godot-mono;
  godot_4-export-templates = godotPackages.godot_4.godot.export-templates;
  godot = godotPackages.godot.godot;
  godot-mono = godotPackages.godot.godot-mono;
  godot-export-templates = godotPackages.godot.godot.export-templates;
```

or the fact that we have confusing attrpaths like `godotPackages.godot.godot`.

The idea is that we've centralised everything in `godotPackages`, and we have an attrset for each version.  Some are aliased, like `godot -> godot_4 -> godot_4_4`.  Each version has `godot` and `godot-mono`.  The export templates are currently in `passthru.export-templates` for each godot package.

Maybe we don't need top level attributes for all of these?

I always struggle with how to structure things like this, so feedback is welcome.  There are so many different examples of things with multiple versions structured in different ways:

- `flutterPackages.v3_27`
- `llvmPackages_17.clang`
- `javaPackages.compiler.openjdk11`
- etc

corngood committed:
godot: add bin/godot symlink

corngood committed:
godot: add project test

corngood commented:
I've cleaned up the top level attributes a bit.  I've also added a test for building and running a test project using the export templates.

I'm pretty happy with this now, so any review would be appreciated.  I have some ideas for further work (e.g. souce-built export template), but there's no reason to do more in this PR.
